src/connection.rs

```diff
@@ -283,8 +286,16 @@ impl<T: Read + Write + ScmSocket> HttpConnection<T> {
                 let line = &self.buffer[*line_start_index..line_end_index];
                 match request.headers.parse_header_line(line) {
                     // If a header is unsupported we ignore it.
-                    Ok(_)
-                    | Err(RequestError::HeaderError(HttpHeaderError::UnsupportedValue(_, _))) => {}
+                    Ok(_) => {
+                        if let Some(limit) = self.maybe_max_len {
+                            if request.headers.content_length() as usize > limit {
+                                return Err(ConnectionError::ParseError(
+                                    RequestError::SizeLimitExceeded,
+                                ));
+                            }
+                        }
```

Comment:
I would move this logic into the part of the connection which handles the transition from parsing headers to parsing body, specifically line 247 in this file. We don't need to check this every single time we parse a new header, we can check it once after we are done with all the headers. In practice, the header section of a request is much smaller than the payload.